Support for vaExportSurfaceHandle

yol opened the pull request:
## Description
Reorganize Vaapi EGL interop so we can also render with vaExportSurfaceHandle

This is WIP, but open for discussion.

TODO:
- [x] Check whether this still works on intel, especially HEVC10
- [x] Modify interop check name to deep color instead of hevc (since it doesn't really check anything hevc-related)
- [x] Find out if this works reliably on AMD due to possibly missing vaSyncSurface support
- [x] Check vaapi and ffmpeg postprocessing/deinterlacing

## Motivation and Context
Hardware decoding with AMD GPUs is only supported with this new method

## How Has This Been Tested?
Run-tested with H.264 and HEVC10 on Linux/Wayland/AMD GPU

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

pkerling committed:
[vaapi] Remove unused attribute querying and warning

pkerling committed:
[vaapi] Request P010 surfaces for 10-bit decoding

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -113,16 +113,13 @@ bool CVaapiTexture::Map(CVaapiRenderPicture *pic)
       glTexParameteri(m_interop.textureTarget, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
       glTexParameteri(m_interop.textureTarget, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
       m_interop.glEGLImageTargetTexture2DOES(m_interop.textureTarget, m_glSurface.eglImageY);
-      glGetIntegerv(GL_IMPLEMENTATION_COLOR_READ_FORMAT, &format);
```
keep this, I use this from time to time

pkerling replied:
Not sure I follow - you use it for debugging? Because the variable clearly isn't used at the moment

FernetMenta replied:
it's there because drivers had issues. currently I only checked in debugger but better would be a check in code if the returned format is correct.

pkerling replied:
Guess I won't touch it for now, but it could definitely use some clean-up

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -187,38 +187,6 @@ bool CVaapiTexture::Map(CVaapiRenderPicture *pic)
 
       break;
     }
-    case VA_FOURCC('B','G','R','A'):
```
keep it

pkerling replied:
You mean 1681a9a completely? As far as I could tell, the resulting imported texture is not used anywhere (at the moment)

FernetMenta replied:
yes, don't touch the existing path and add the new one by implementing a new class. this new class can reside in VaapiEGL.cpp

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.h:
```diff
@@ -59,7 +59,7 @@ class CVaapiTexture
   int m_texWidth = 0;
   int m_texHeight = 0;
 
-protected:
```
keep as is

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -175,21 +308,25 @@ bool CVaapiTexture::Map(CVaapiRenderPicture *pic)
+  if (m_vaapiPic)
+    return true;
+  
+  if (!MapEsh(pic) && !MapImage(pic))
```
I don't like this. keep it simple, we either do this or that. we know at the very biginning with path we'll take.

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -201,29 +338,39 @@ void CVaapiTexture::Unmap()
   if (!m_vaapiPic)
     return;
 
-  if (m_glSurface.vaImage.image_id == VA_INVALID_ID)
```
keep this method as is and add an addition one for the new mapping

FernetMenta commented:
first create a new class for the new mapping method. the renderer should instanciate the right one on configure. later we can promote common code to a common super class.

pkerling commented:
Where do we want to decide on the method to use? Theoretically, we do the interop test at app start anyway, and already know then if we can use `vaExportSurfaceHandle` or not. Or is is it better to wait until starting playback and decide then?

pkerling commented:
> first create a new class for the new mapping method. the renderer should instanciate the right one on configure. later we can promote common code to a common super class.

OK I'll try to split it to two classes

FernetMenta commented:
in current code there is nothing to split. the right word is add :)

pkerling commented:
Wouldn't it be easier to go for a common base class from the start? Or at least an interface for Map/Unmap, otherwise I have to special-case everything in the decoder.

FernetMenta commented:
Yes, a common interface and these should be pointers:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVAAPIGL.h#L66

pkerling commented:
this now has minimal changes to the vaapi1 code path

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -200,43 +199,20 @@ bool CVAAPIContext::CreateContext()
   m_profileCount = 0;
 
-  int max_attributes = vaMaxNumDisplayAttributes(m_display);
-  m_attributes = new VADisplayAttribute[max_attributes];
```
why do you drop this?

pkerling replied:
Same as everything else I dropped previously - not used anywhere. And I don't think it's good to warn about something not being there that's not being used in either case. Or did you want it just for the log output?

FernetMenta replied:
the log has become useless, right? ok to drop

pkerling replied:
Yeah mesa does not export any display attributes and we actually don't use any

FernetMenta commented on xbmc/utils/EGLUtils.cpp:
```diff
@@ -33,6 +33,24 @@ std::set<std::string> CEGLUtils::GetClientExtensions()
   return result;
 }
 
+std::set<std::string> CEGLUtils::GetExtensions(EGLDisplay eglDisplay)
```
I don't see those used anywhere?

pkerling replied:
https://github.com/xbmc/xbmc/pull/13333/files/4a0785799de98b02f1ea86f958de36b70b27531b#diff-f4632009b5b4bdf3f11d1aaf091cee66R494

FernetMenta replied:
disregard :)

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVAAPIGL.cpp:
```diff
@@ -77,7 +77,9 @@ bool CRendererVAAPI::Configure(const VideoPicture &picture, float fps, unsigned
   {
-    tex.Init(interop);
+    // FIXME How do we know which texture type to create without running the interop test every time?
+    tex.reset(VAAPI::CVaapiTexture::CreateTexture(CRendererVAAPI::m_pWinSystem->GetVADisplay(), CRendererVAAPI::m_pWinSystem->GetEGLDisplay()));
```
don't call a static. here is the location where we either instantiate one class or the other.

pkerling replied:
Yeah I don't really like it like this either (thus the FIXME), but do you have a suggestion about who should decide which texture interop to create? I don't think it concerns the renderer directly (as long as the textures are mapped in GL, it doesn't matter how) so it shouldn't do it itself.

FernetMenta replied:
CRendererVAAPI takes the decision here. There can be non-EGL mappings like the old putSurface for GLX.

pkerling replied:
Sure, but for the problem at hand both VAAPI1 and VAAPI2 map to GL, so the renderer does not care which it is. It doesn't need to do anything differently.

FernetMenta replied:
putSurface for GLX would also map to GL. But this class certainly won't reside in a file named EGL. here is a good location.

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVAAPIGL.cpp:
```diff
@@ -77,7 +77,9 @@ bool CRendererVAAPI::Configure(const VideoPicture &picture, float fps, unsigned
   for (auto &tex : m_vaapiTextures)
   {
-    tex.Init(interop);
+    // FIXME How do we know which texture type to create without running the interop test every time?
```
yes, we need to run the test. a user may have switched the app from one display to another.

pkerling replied:
So what's the point of having the interop test run at app start at all? Shouldn't we register the vaapi renderer always then?

FernetMenta replied:
> So what's the point of having the interop test run at app start at all?

test if vaapi decoder can handle 10bit. X11 registers it every time a new context gets created.

pkerling replied:
Ah, if X11 does it every time then it makes sense. I'll try to get it sorted out.

pkerling commented:
@FernetMenta Interop test moved to renderer as requested

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVAAPIGL.cpp:
```diff
@@ -75,9 +75,31 @@ bool CRendererVAAPI::Configure(const VideoPicture &picture, float fps, unsigned
 
+  bool useVaapi2{};
+  bool general{}, hevc{};
+#if KODI_HAVE_VAAPI2TEXTURE
```
remove the ifdef here, instead make CVaapi2Texture::TestInterop return false, if !KODI_HAVE_VAAPI2TEXTURE

pkerling replied:
OK, that would mean ifdef-ing every function in CVaapi2Texture though? Either in the header or in the implementation.

FernetMenta replied:
you could define a kind of stub class for this case. then no ifdefs are need for the actual class.

pkerling replied:
Mh don't really like that idea, then people will forget to keep the stub in sync and build will break on older libva. I've ifdef-ed the places that need it, there weren't that many after all.

FernetMenta replied:
I won't forget and I maintain this code

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.h:
```diff
@@ -31,6 +33,12 @@
+#include "guilib/Geometry.h"
+#include "utils/posix/FileHandle.h"
+
+#define KODI_HAVE_VAAPI2TEXTURE VA_CHECK_VERSION(1,1,0)
```
I see no benefit of this marco, keeping VA_CHECK_VERSION(1,1,0) saves the time required for looking up the meaning of this macro

pkerling replied:
It's for having the conditions for enabling Vaapi2Texture coherent in the code. We don't even really know what VAAPI version this will be in at the moment.

FernetMenta replied:
ok, but drop the KODI prefix. that pretends that availability is defiend by our code.

pkerling replied:
done

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -24,15 +24,35 @@
+  general = hevc = false;
+#if KODI_HAVE_VAAPI2TEXTURE
+  CVaapi2Texture::TestInterop(vaDpy, eglDisplay, general, hevc);
+  CLog::Log(LOGDEBUG, "Vaapi2 EGL interop test results: general %s, hevc %s", general ? "yes" : "no", hevc ? "yes" : "no");
```
those logs just generate log spam. drop them here. better log at the location it gets registered.

pkerling replied:
Where it's registered we don't know whether it's vaapi1 or vaapi2 (since it does not matter for registering)

FernetMenta replied:
same rule as for configure. don't mix v1 and v2. register is done here:
CRendererVAAPI::Register, for all methods.

pkerling replied:
done

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -445,8 +468,279 @@ bool CVaapiTexture::TestInteropHevc(VADisplay vaDpy, EGLDisplay eglDisplay)
+        return false;
+    }
+
+    EGLint attribs[13 + 4 /* modifiers */] = {
```
drop those ugly comments interweaved with code. if required add at end of line

pkerling replied:
done

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -445,8 +468,279 @@ bool CVaapiTexture::TestInteropHevc(VADisplay vaDpy, EGLDisplay eglDisplay)
+      return false;
+    }
+
+    texture->glTexture = ImportImageToTexture(texture->eglImage);
```
never create a method that is only called once. better add those 6 lines of code here.

pkerling replied:
It will be called twice once it is moved to the base class and `Vaapi1Texture` also uses it :-)

FernetMenta replied:
never create methods that are not needed now but maybe in the future

pkerling replied:
Why in the future? I would have done it as part of this PR once the Vaapi2 stuff is OK.

Also, splitting code into appropriately named functions (even if they are not reused) is actually good practice because it makes the code more self-describing and easier to read/test/debug if each function only does one thing. I see that we are very far from that in many parts of the Kodi code, but why should we actively work against it in new code?

FernetMenta replied:
I don't see the function being used more that once in this diff. hence drop it

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -445,8 +456,282 @@ bool CVaapiTexture::TestInteropHevc(VADisplay vaDpy, EGLDisplay eglDisplay)
+  hevc = false;
+
+  int major_version, minor_version;
+  if (vaInitialize(vaDpy, &major_version, &minor_version) != VA_STATUS_SUCCESS)
```
this has to be moved up a level or it will be callen on each configure

pkerling replied:
Hm. I wonder, should we just keep an initialized VADisplay around in the win system instead of returning a new one in `GetVADisplay()` every time?

FernetMenta replied:
yes, we can keep it in CVaapiProxy

FernetMenta replied:
nope, see my other comment. VADisplay is passed in CVaapiRenderPicture.

pkerling replied:
moved it to CVaapiProxy now (only for Wayland to test), because it still saves reinitializations each time you start a movie and you don't have to worry in TestInterop about whether you have to initialize the display or not. can move it back though.

FernetMenta replied:
this is wrong, VADisplay is carried in VaapiPresentPicture. this can be used in configure

pkerling replied:
VADisplay form VaapiPresentPicture is being used, see https://github.com/xbmc/xbmc/pull/13333/commits/015867047e3ae2b9b63ac343adecfd7dc5b329b6#diff-8869e60284b5a072aa8ffd8e1e7f0fb4R88

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -445,8 +456,282 @@ bool CVaapiTexture::TestInteropHevc(VADisplay vaDpy, EGLDisplay eglDisplay)
+    hevc = TestEsh(vaDpy, eglDisplay, VA_RT_FORMAT_YUV420_10BPP, VA_FOURCC_P010);
+  }
+
+  vaTerminate(vaDpy);
```
that is even more critical here

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVAAPIGL.cpp:
```diff
@@ -75,9 +83,20 @@ bool CRendererVAAPI::Configure(const VideoPicture &picture, float fps, unsigned
   interop.eglDisplay = CRendererVAAPI::m_pWinSystem->GetEGLDisplay();
 
+  bool useVaapi2{}, hevc{};
+  VAAPI::CVaapi2Texture::TestInterop(CRendererVAAPI::m_pWinSystem->GetVADisplay(), CRendererVAAPI::m_pWinSystem->GetEGLDisplay(), useVaapi2, hevc);
```
VADisplay in carried in CVaapiRenderPicture

pkerling replied:
fixed

pkerling commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -152,9 +152,18 @@ void CVAAPIContext::SetValidDRMVaDisplayFromRenderNode()
+      if (CheckSuccess(vaInitialize(display, &major_version, &minor_version)))
+      {
+        m_renderNodeFD = fd;
+        m_display = display;
```
Note: This is missing a vaTerminate now. Will fix if we keep going this way.

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -185,58 +193,24 @@ bool CVAAPIContext::CreateContext()
     return false;
   }
 
-  int major_version, minor_version;
```
don't change this. I don't want vaapi to be initialized when not used

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.h:
```diff
@@ -31,6 +33,10 @@
 #include <va/va.h>
+#include <va/va_drmcommon.h>
+
+#include "guilib/Geometry.h"
```
this dependecy to guilib is kind of ugly. eventually VP should be used transcoding on a headless system without a gui. I don't see much benefit in using this CSizeInt class here.

pkerling replied:
ofc it's not strictly necessary - for me the benefit here is that they encapsulate the return value in one atomic unit. Otherwise you'd have to have two functions for width and height, but 100% of the time you need both anyway. Or pass references as out arguments but that's also kind of ugly. 
Actually, I don't see why CSizeInt etc. should be in guilib at all. Going forward I think it would actually be good to use these classes where appropriate since they encapsulate related data and support useful operations with them. They are not really related or limited to the GUI at all. I use them a lot in the Wayland code already, which is also not quite GUI. Could we move them out of guilib?

FernetMenta replied:
I don't mind if you move it

pkerling replied:
Is it OK if I make a separate PR for that afterwards? It will be a search-replace job and should not be mixed with the vaapi changes here.

FernetMenta replied:
ok, if you are going to do this in the near future

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVAAPIGL.cpp:
```diff
@@ -75,9 +92,20 @@ bool CRendererVAAPI::Configure(const VideoPicture &picture, float fps, unsigned
   interop.eglDisplay = CRendererVAAPI::m_pWinSystem->GetEGLDisplay();
 
+  bool useVaapi2{}, hevc{};
+  VAAPI::CVaapi2Texture::TestInterop(pic->vadsp, CRendererVAAPI::m_pWinSystem->GetEGLDisplay(), useVaapi2, hevc);
```
here we want to test either 10bit or 8bit. no need for testing both

pkerling replied:
done

pkerling commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -445,8 +444,273 @@ bool CVaapiTexture::TestInteropHevc(VADisplay vaDpy, EGLDisplay eglDisplay)
+  if (!m_vaapiPic)
+    return;
+
+  Unmap(m_y);
```
I can rename it if you find it misleading. The benefit of having it is not having to repeat code.

FernetMenta replied:
the code is easier to read if the function is expanded here. this I rank higher than saving 2 lines of code.
the desire of coders to compact code in early stages is one reason for this spaghetti code we've got. next step is often to create a parameter with a default value when reuse of a method does not exactly fit.
you won't find a suitable name because your primary motivation is grouping random parts of code for reuse.

pkerling replied:
> the desire of coders to compact code in early stages is one reason for this spaghetti code we've got.

I don't see how this is spaghetti code, and I'm not sure whether you are suggesting it is or just fear that it will turn into such later?

> you won't find a suitable name because your primary motivation is grouping random parts of code for reuse.

It unmaps a texture that was mapped into EGL+GL from DRM. It's a decomposition of the `CVaapiTexture::Unmap` function (that unmaps two textures) into one unmap step for each of the two textures. If `Unmap` is not a fitting name, then `CVaapiTexture::Unmap` is already unfitting.

In principle, I get your argument. Grouping "random parts of code" into functions isn't helpful. However, I don't see why you think it is random (i.e. it does not have one clearly defined purpose) here.

FernetMenta replied:
> I don't see how this is spaghetti code, and I'm not sure whether you are suggesting it is or just fear that it will turn into such later?

I was referring to kodi code in general and what I have seen during the last couple of years. 
Since you wrote this code, you know what "unmap" does. For somebody who may study this code in the future it means that she/he must navigate to this function in order to understand what it does.

FernetMenta replied:
btw: don't forget to drop those methods

pkerling commented:
I checked the postproc stuff and the ffmpeg won't work any more because it relies on `vaDeriveImage`. Theoretically, it would be possible to modify it to use `vaExportSurfaceHandle` and then map the dmabuf handle and copy the data from there. Is `CFFmpegPostproc` used a lot and do we need to support it on AMD?

FernetMenta commented:
CFFmpegPostproc is used for bob and yadif deinterlacing.

pkerling commented:
Is this something users want/need? Or is VAAPI deinterlacing usually sufficient?

Also: Why do we need to copy the image to system RAM for bob deinterlacing?

fritsch commented:
From my pov (my personal pov), this code was highly needed back at the time when VPP did not properly work on intel. We copied the surfaces back with SSE4 and did SW deinterlacing on the CPU or RenderBob. This needs alone for the CPU copy a huge amount of CPU cycles.

Since this time a lot has happened:
Intel fixed MADI on SNB and IVB, HSW and later can use MCDI.

In short: I don't think we should still maintain this sw code anymore.

FernetMenta commented:
In general VAAPI deinterlacing is ok, at least on Intel. Nevertheless, we keep the other methods.

> Also: Why do we need to copy the image to system RAM for bob deinterlacing?

our bob deinterlacing method requires this. does not mean that it can't be done differently.

FernetMenta commented:
> in short: I don't think we should still maintain this sw code anymore.

I don't agree. It's at least still a good benchmark for vpp deinterlacing methods.

fritsch commented:
@FernetMenta currently it's broken. If you disable "Prefer VAAPI render method" - you will say hello to a green screen only. Therefore I asumed it is neither maintained and as you did not know yet also not maintained anymore.

FernetMenta commented:
Nobody brought to my attention. I am not aware of a bug report.

fritsch commented:
That's exactly what I tried to say: It's not used, by no one. I tried it after reading the discussion here. First time after 1 year or something.

lrusak commented:
just tested this and it seems I cannot play 10bit hevc on intel anymore. I just get a green screen and in the log I see:
```
11:31:21.789 T:139846154621888   ERROR: Failed to import VA DRM surface into EGL image (EGL error 12291)
11:31:21.909 T:139846122870528   ERROR: Previous line repeats 2 times.
```
full log http://termbin.com/ri9a

fritsch commented:
Mmh:
https://github.com/01org/intel-vaapi-driver/commit/106d1eb730210b498c1f9fca9f4f5c3df82d056d#diff-5ac9c57cd3d2ce7d8458ba756c3582c0R6522

2018-01-20 20:36 GMT+01:00 Lukas Rusak <notifications@github.com>:

> just tested this and it seems I cannot play 10bit hevc on intel anymore. I
> just get a green screen and in the log I see:
>
> 11:31:21.789 T:139846154621888   ERROR: Failed to import VA DRM surface into EGL image (EGL error 12291)
> 11:31:21.909 T:139846122870528   ERROR: Previous line repeats 2 times.
>
> full log http://termbin.com/ri9a
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/13333#issuecomment-359196519>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABCfHSSYdIcD-D4MfgqZS4sEMsYLkxmZks5tMkA2gaJpZM4RWSAg>
> .
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

lrusak commented:
>https://github.com/01org/intel-vaapi-driver/commit/106d1eb730210b498c1f9fca9f4f5c3df82d056d#diff-5ac9c57cd3d2ce7d8458ba756c3582c0R6522

:disappointed:

fritsch commented:
@lrusak that's perfectly fine. See how the old code was done. And let's wait for @pkerling on how to properly handle it for both AMD and intel now.

pkerling commented:
@lrusak I'm in the process of getting it to run on my NUC atm, then I can hopefully fix it. Takes more time than I thought it would :-)

pkerling commented:
@lrusak 01org/intel-vaapi-driver#342

pkerling committed:
[vaapi] Replace VAProfileH264Baseline with ConstrainedBaseline

Baseline profile support in VAAPI is deprecated and was never
implemented in any driver.

pkerling committed:
[vaapi] Be more explicit about point of failure in CVppPostproc error messages

pkerling committed:
[vaapi] Hand picture_width and picture_height to vaCreateContext

Mesa vaapi drivers refuse to create the context without this
information.

pkerling committed:
Add EGL utilities for extensions and attributes

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
[vaapi] Make CVaapiTexture into an interface, factor out vaapi1

Preparation work for adding CVaapi2Texture

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
[vaapi] Add vaExportSurfaceHandle support

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
[vaapi] Rename hevc flag to deepColor

The flag actually has nothing to do with HEVC,
but checks for > 8 bit import support.

pkerling commented:
Rebased.

@FernetMenta 064247699d909d57a46a7c8a3d6066a29a7a91f3 fixes deinterlace disabling on the surface, but it causes some problems. Sometimes it crashes, sometimes I get decoder timeouts etc. Would you be willing to look into it? I'm not really familiar with the state machines etc.

FernetMenta commented:
@pkerling this looks wrong. deinterlacing flag can toggle or player can disable it temporarily to catch up if late. you don't want a heavy operation like deinit/init prosproc for those cases.

pkerling commented:
OK, so what should we do instead? Temporarily bypass the filters?

FernetMenta commented:
IMO they are correct. vaapi postproc can just do nothing if no mode requested.

pkerling commented:
> IMO they are correct. vaapi postproc can just do nothing if no mode requested.

What do you mean with "they?" Current code is not correct. If you try to disable deinterlacing in the interface, nothing will happen. We need to change something.

FernetMenta commented:
now I remember. I'll fix it

FernetMenta commented:
fixed by https://github.com/xbmc/xbmc/pull/13518

pkerling commented:
Switching deinterlace on/off works, but I still get timeouts and crashes.

pkerling commented:
OK, seems to be a mesa bug: https://bugs.freedesktop.org/show_bug.cgi?id=105145

pkerling committed:
[vaapi] Print profile string on vaapi2

pkerling committed:
[vaapi] Support skipping deinterlace filter in CVppPostproc

pkerling committed:
[vaapi] Adjust postproc choice for mesa limitations

Videos <=1080p will now use CVppPostproc even when no deinterlacing
is performed, so that it can be dynamically switched on/off.

pkerling commented:
@FernetMenta this should be about it now. Maybe there'll be a flag to tell mesa to frame-decode by default so it doesn't have to weave the first few frames, but this can go into a separate PR if it's done. So you can review the current state when you have time :-) I'll update the GLES renderer after your approval.

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -2358,9 +2358,35 @@ bool CVppPostproc::PreInit(CVaapiConfig &config, SDiMethods *methods)
+{
+  if (method == m_vppMethod)
+  {
+    // Nothing to do
```
no need to comment the obvious

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -2493,9 +2514,10 @@ bool CVppPostproc::Filter(CVaapiProcessedPicture &outPic)
   }
   outPic.DVDPic.SetParams(it->DVDPic);
 
+  const bool skipDeinterlacer = ((m_vppMethod == VS_INTERLACEMETHOD_NONE) || (outPic.DVDPic.iFlags & DVD_CODEC_CTRL_SKIPDEINT));
```
why did you change this? you broke behaviour. SKIPDEINT skips the 2nd cycle but not entire deinterlacing

pkerling replied:
Unintentional, will take a look

pkerling replied:
OK should be fine the way it was even with handling `VS_INTERLACEMETHOD_NONE`. I probably misread something.

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1920,94 +1920,92 @@ void COutput::InitCycle()
   }
-  if (!m_pp) // fallback
+
+  switch (method)
```
I don't like this change at all. I want to keep deinterlacing and progressive in separate paths like it was before.

pkerling replied:
Then you have to tell me where I should differentiate. I assume c&p-ing the same code to a if and else branch will not make you happy :-)

FernetMenta replied:
> I assume c&p-ing the same code to a if and else branch will not make you happy :-)

you might be wrong. I don't mind some duplicate lines of code if the design is clear.

pkerling commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1928,92 +1920,142 @@ void COutput::InitCycle()
       {
+        CLog::Log(LOGINFO, LOGVIDEO, "VAAPI output: Initializing skip postproc");
         m_pp = new CSkipPostproc();
-        m_config.stats->SetVpp(true);
```
Actually - what meaning does `SetVpp` have? Should it be true for SkipPostproc?

FernetMenta replied:
ok, does not have any function atm

pkerling commented:
@FernetMenta I gave it my best try ;-)

FernetMenta commented:
please squash the fixup commits. then I can review the changes separately

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1939,43 +1931,57 @@ void COutput::InitCycle()
-          m_config.processInfo->SetVideoDeintMethod("vaapi-madi");
-        else if (method == VS_INTERLACEMETHOD_VAAPI_MACI)
-          m_config.processInfo->SetVideoDeintMethod("vaapi-mcdi");
+        switch (method)
```
I mean the entire diff is already ver big and cosmetics should be done in a separate PR

pkerling commented:
Done. I initially left them so you knew what I changed compared to the last iteration.

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1931,21 +1931,23 @@ void COutput::InitCycle()
       delete m_pp;
       m_pp = NULL;
       DropVppProcessedPictures();
-      m_config.processInfo->SetVideoDeintMethod("unknown");
```
why do you drop this?

pkerling replied:
It will be overwritten by either the correct implementation or `skip` as fallback anyway (which I added), so it is not needed anymore.

FernetMenta replied:
it reflects the state at this time. if anything goes wrong later, it won't get overwritten
keep it

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1977,19 +1980,37 @@ void COutput::InitCycle()
+      // and use for direct export, so we run into trouble if we or the user want to switch
+      // deinterlacing on/off mid-stream.
+      // See also: https://bugs.freedesktop.org/show_bug.cgi?id=105145
+      const bool alwaysInsertVpp = (m_config.vidWidth * m_config.vidHeight) <= (1920 * 1080);
```
we don't do this on Intel

pkerling replied:
Well then I have to make it a driver-specific quirk. Is that OK?

FernetMenta replied:
yes, check on open and store in config for later use

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -2784,11 +2784,6 @@ bool CFFmpegPostproc::PreInit(CVaapiConfig &config, SDiMethods *methods)
 {
   m_config = config;
   bool use_filter = true;
-  if (!m_dllSSE4.Load())
```
why?

pkerling replied:
As the commit message says: This function will always return false on mesa, in which case there is no point in loading `m_dllSSE4`. So I moved the loading further down when we know whether we actually need it or not.

FernetMenta replied:
ok

pkerling commented:
jenkins build this please

pkerling commented:
Jenkins is green: http://jenkins.kodi.tv/job/BuildMulti-PR/7273/

@FernetMenta from my side this is good to go

FernetMenta commented:
I want to runtime test this myself before merge

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1949,71 +1943,98 @@ void COutput::InitCycle()
       }
     }
+
+    if (method == VS_INTERLACEMETHOD_DEINTERLACE)
```
this block is wrong here. we don't want to set the method on every cycle even if nothing changed

pkerling replied:
Code will get more complicated, but sure

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVAAPIGL.cpp:
```diff
@@ -185,14 +212,15 @@ bool CRendererVAAPI::UploadTexture(int index)
 
-  m_vaapiTextures[index].Map(pic);
-  m_buffers[index].m_srcTextureBits = m_vaapiTextures[index].GetBits();
+  m_buffers[index].m_srcTextureBits = m_vaapiTextures[index]->GetBits();
```
this is wrong. why did you change the order? bits is set in Map()

pkerling replied:
I vaguely remember this being introduced in a rebase, so probably rebase screwup

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -473,8 +472,284 @@ bool CVaapiTexture::TestInteropHevc(VADisplay vaDpy, EGLDisplay eglDisplay)
+
+    case VA_FOURCC_P010:
+    case VA_FOURCC_P016:
+      m_bits = 16;
```
this is not correct. keep it the same as Texture1

pkerling replied:
OK. Then maybe you could document what the bits value is being used for ;-)

pkerling replied:
Setting this to 10 here breaks hevc10 display for me, everything goes pink

pkerling replied:
Ah sorry, didn't see the overwrite further down. Maybe you should remove this line: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp#L158 It's not being used and quite confusing

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1949,71 +1943,98 @@ void COutput::InitCycle()
       if (method == VS_INTERLACEMETHOD_DEINTERLACE ||
           method == VS_INTERLACEMETHOD_RENDER_BOB)
       {
+        CLog::Log(LOGINFO, LOGVIDEO, "VAAPI output: Initializing ffmpeg postproc");
```
this should not be LOGINFO, make it component logging for video as the other logs here

pkerling replied:
It is LOGVIDEO

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1949,71 +1943,98 @@ void COutput::InitCycle()
       }
       else
       {
+        CLog::Log(LOGINFO, LOGVIDEO, "VAAPI output: Initializing vaapi postproc");
```
same here

pkerling replied:
It is LOGVIDEO

FernetMenta replied:
LOGINFO -> LOGDEBUG

pkerling committed:
[vaapi] Soften FFmpegPostproc failure log level, delay dllSSE4 load

FFmpegPostproc failing is expected on mesa VAAPI drivers since we do
not support in conjunction with vaExportSurfaceHandle and those drivers
cannot do vaDeriveImage.

pkerling commented:
hm fast-forward seems to be broken on AMD when deinterlacing, I'll investigate

pkerling commented:
OK I've found part of the problem. It happens also on master and with sw-decoding. To reproduce:

1. Open an interlaced video and set deinterlace method to none
1. Stop playback
1. Open the same file again (so it starts without deinterlacing)
1. Immediately start fast-forwarding
1. At around 5 seconds, renderer reconfigures and then all frames get dropped here: https://github.com/xbmc/xbmc/blob/0fcc22546430f3f57f2f258c33c1c720c61e9398/xbmc/cores/VideoPlayer/VideoPlayerVideo.cpp#L902

There might also be other paths to trigger this bug, but this works reliably.

pkerling commented:
https://bugs.freedesktop.org/show_bug.cgi?id=105368

pkerling commented:
Grr oversight, thanks for catching. I had log statements pretty much everywhere for debugging and this one slipped through.

FernetMenta commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp:
```diff
@@ -1947,85 +1942,125 @@ void COutput::InitCycle()
       }
+
+      if (method == VS_INTERLACEMETHOD_DEINTERLACE)
+        m_config.processInfo->SetVideoDeintMethod("yadif");
```
Setting deinit method can be done by Cpostproc. It has access to config. Then resurrecting m_currentDiMethod is not needed

pkerling commented:
Then I'll just hit the button...

Thanks for thoroughly reviewing this. I know we have very differing opinions about code organization, but I do appreciate the time you are taking to look over all of this stuff.

fritsch commented:
@pkerling @FernetMenta thx very much!

@pkerling Could you please document here which version of e.g. libva / mesa / additional patches are needed, cause these question will arrise very soon.

pkerling commented:
In summary:

- libva 2.1
- mesa master or 17.3 with this patch: https://lists.freedesktop.org/archives/mesa-dev/2017-December/178971.html

ffmpeg does not need any patches.

fritsch commented:
@lrusak @MilhouseVH ^^ as you are running libva 2.1 already, mind picking above patch on next LE nightly builds? That way we can start to test this on various AMD GPUs.

garbear commented:
I'm getting the following compiler error on Ubuntu 16.10:

```
VaapiEGL.cpp:454:35: error: ‘DRM_FORMAT_GR1616’ was not declared in this scope
         EGL_LINUX_DRM_FOURCC_EXT, DRM_FORMAT_GR1616,
                                   ^~~~~~~~~~~~~~~~~
```

My version of installed libdrm is 2.4.70-1.

Rechi commented:
libdrm 2.4.82 or newer is required
you can get 2.4.83 from xenial-updates

garbear commented:
Adding
```
deb http://archive.ubuntu.com/ubuntu/ xenial-updates main restricted
deb-src http://archive.ubuntu.com/ubuntu/ xenial-updates main restricted
```

to `/etc/apt/sources.list` and running `apt-get install libdrm-dev` fixed it. thanks

FernetMenta commented:
@pkerling  this was the reason why it was fourcc_code('G', 'R', '3', '2')
please change it back

pkerling commented:
Why? 16.10 is EOL for almost a year and we don't support it any more even for Krypton.

FernetMenta commented:
it has to compile on 16.04

pkerling commented:
It does. As @rechi said 16.04 has libdrm 2.4.83 in updates.

FernetMenta commented:
ok

bkuhls commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp:
```diff
@@ -475,3 +478,274 @@ bool CVaapi1Texture::TestInteropHevc(VADisplay vaDpy, EGLDisplay eglDisplay)
+    auto const& layer = surface.layers[layerNo];
+    if (layer.num_planes != 1)
+    {
+      CLog::LogFunction(LOGDEBUG, "CVaapi2Texture::Map", "DRM-exported layer has %d planes - only 1 supported", layer.num_planes);
```
Can this restriction be removed from Kodi?
Video playback on Intel Rocketlake CPUs is broken atm ( https://github.com/xbmc/xbmc/issues/19820 ).

mpv handles multiple planes just well and has no problems with video playback on Intel Rocketlake CPUs with this code:
https://github.com/mpv-player/mpv/commit/05cb8d28afd8881667c3dea297b182a2ecdd91c8#diff-38c39325f6dad76d975c9eee56dc8a14487c425bae13c3f0482f7fe0fbce2fbbR363

yol replied:
Well the API for DRMPRIME external buffer descriptors say that when calling vaExportSurfaceHandle with VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME_2 and VA_EXPORT_SURFACE_SEPARATE_LAYERS, each layer will contain exactly one plane, so that's why it is handled like this here. If Intel broke this, I guess it's their choice? Considering they're very likely also the ones that originally specified it like this. Anyway, I would be open to adding a quirk if it's not much effort (i.e., it "just works" if we add the other plane properties). PRs welcome.

And please don't necro three year old PRs with "reviews" :-)
